TST: Add new tests for array coercion

seberg opened the pull request:
Our current array-coercion tests are not all that bad, however, they are partially scattered
around.
These adds two main (extensive) tests. First that *our* scalars (not python scalars!) behave identically when assigned, or coerced when compared to normal casting.

Second, it adds some additional tests for nested arrays.

The main changes are that our scalars (with the one exception of void, because it is both structured and not structures) always behave the same.

There are two larger "change" that is a bit less obvious:
   * `np.array([np.float64("nan")], dtype=int)` and some other things (identical to the datetime paths here) would use `float(item)` or `int(item)` and then fail for NaN... For *our* scalars, in my branch these do the same thing as casting. Which is currently succeed.
  *  Coercing float128 to complex256, I believe used to lose precision, it will not lose precision now, since it uses the normal casting machinery.

To be clear, the handling of Python scalars does *not* change in my branch. There are a couple of potential corner cases to map out, but please do not let this stop this here. It is probably easier to expand in a second PR anyway.

 * Subclasses of python scalars are unmodified by my changes in most cases, because they are considered "unknown" (except during dtype discovery).


----

I am open to reorganizing this a bit, but its pretty tedious and I am not sure its worth optimizing the code for not repeating patterns much, considering these are tests...

seberg committed:
ENH: Add traceback-skip to `assert_array_compare`

seberg committed:
TST: Add tests for arraycoercion

These tests have many xfails (some technically maybe not correct),
which describe in detail what will change when merging the
array-coercion changes; since all of those xfails are going
to be removed.

seberg committed:
TST: Add test for empty sequences

seberg committed:
TST: Add tests for bad self-mutating sequence inputs to np.array

These should never happen and rightly should lead to undefined
behaviour (and preferably errors), but they excercise some more
tricky code branches and should probably not crash.

seberg committed:
Simplify the "all scalars" logic (hardcode) and fix complex cases

seberg commented:
@mattip do you know from the top of the head why `int(np.complex128(3))` fails on PyPy?  Is it because of the ComplexWarning making problems?

Its not an issue, I will just xfail the test for now on pypy I think. Just curious.  It will probably work on my branch anyway.

seberg committed:
MAINT: Some cleanup, and xfail pypy

PyPy seems to have issues with int(numpy_complex), maybe because
it gives a warning during conversion (python does not define it).
So simply mark it as xfail, it should work in my branch.
Also some smaller cleanups.

mattip commented:
> Why int(np.complex128(3)) fails on PyPy? 

Calling `int(np.complex128(3)) ` fails, but directly calling `__int__` succeeds. Strange since the internal PyPy code for `int` seems to lookup and call `__int__`.

```
$ pypy3 -Walways
Python 3.6.9 (2ad108f17bdb, Apr 07 2020, 02:59:05)
[PyPy 7.3.1 with GCC 7.3.1 20180303 (Red Hat 7.3.1-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>> import numpy as np
>>>> int(np.complex128(3)) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert complex to int
>>>> np.complex64(3)
(3+0j)
>>>> np.complex64(3).__int__()
__main__:1: ComplexWarning: Casting complex values to real discards the imaginary part
3
>>>> np.complex128(3).__int__()
__main__:1: ComplexWarning: Casting complex values to real discards the imaginary part
3
```

mattip commented:
This is a PyPy bug in subclasses of complex, now fixed and will be part of the next release. Thanks for pointing it out.

charris commented:
Hmm, running the tests with `-vv` isn't as informative as might be wished, the parametrized values are simply numbered so that it is difficult to tell what failed. Let me think about this for a while...

seberg commented:
I think there was some way to name the parameters? have to look up how it worked though...

charris commented:
> I think there was some way to name the parameters?

I couldn't think of good names :) It also wouldn't fit nicely into your generator approach. Maybe if the yielded values could simply have a comment `scalar0, scalar1, ...` to mark them. But one would still need to look at the code to see what failed and it would require more maintenance effort.

charris commented:
Could have the generator functions return a (named)tuple of two lists, one of names and the other of values. could then pass that as `scalar.names, scalar.values` or use it to make a pytest fixture. Alternatively, return a list of tuples and unzip them
```
>>> l = [(1,2), (3,4), (8,9)]
>>> list(zip(*l))
[(1, 3, 8), (2, 4, 9)]
```

seberg commented:
OK, want to look into a few new tests (and clean up the fallout that always comes if you test edge-cases of edge-cases) and probably a couple of small, additional benchmarks just to document the changes.

The named tuple I know roughly how to do. I am not even sure the yields are perfect, I could also just make lists, but I doubt that would help much.  Some custom fixture, or hypothesis like things also could work I guess, but have to find out how to do that first...

seberg committed:
TST: Add some further test, re-add missing rational

seberg committed:
MAINT: Make rationals optional, because they fail some tests currently

The failures will go away, since this must again be related to
implemetning int() and float().

seberg commented on numpy/conftest.py:
```diff
@@ -89,6 +89,6 @@ def check_fpu_mode(request):
-@pytest.fixture(autouse=True)
-def add_np(doctest_namespace):
-    doctest_namespace['np'] = numpy
+#@pytest.fixture(autouse=True)
+#def add_np(doctest_namespace):
+#    doctest_namespace['np'] = numpy
```
```suggestion
@pytest.fixture(autouse=True)
def add_np(doctest_namespace):
    doctest_namespace['np'] = numpy
```
Woops, that change was accidentally commit (its necessary for pytest-leaks unfortunately).

seberg committed:
Update numpy/conftest.py

charris commented:
Named tuples don't look useful for this application. I think lists are an option.

anirudh2290 commented on numpy/core/tests/test_array_coercion.py:
```diff
@@ -0,0 +1,524 @@
+        assert arr.shape == (1,)
+
+    def test_char_special_case(self):
+        arr = np.array("string", dtype="c")
```
is this operation the same as ```arr = np.array("string", dtype="S1")``` ?
if i use dtype as S1 it gives a different result:
```
>>> arr = np.array("string", dtype="S1")
>>> arr
array(b's', dtype='|S1')
```

seberg replied:
Yeah, its weird, `"c"` is the same as `"S1"`, but with the special case that `arr.dtype.char` is `c` and that here the result gives 6 individual characters instead of the `"string"`.  We should get rid of it some of these days IMO.

charris replied:
IIRC, we have been planning to get rid of `c` for many years. It may be used somewhere, though. I think it was originally for data input for old applications.

seberg replied:
Yeah, we definitely should get rid of it, I just didn't feel like looking into it too deeply, because it seemed just as easy to add the special case for now.  I think `c` is actually partially deprecated on the C-API side, but not on the python side...

anirudh2290 commented on numpy/core/tests/test_array_coercion.py:
```diff
@@ -0,0 +1,524 @@
+            nested = [nested]
+
+        arr = np.array(nested, dtype='c')
+        assert arr.shape == (1,) * (np.MAXDIMS - 1) + (6,)
```
This test will also fail if i replace dtype='c' with dtype='S1'

anirudh2290 commented on numpy/core/tests/test_array_coercion.py:
```diff
@@ -0,0 +1,524 @@
+    yield np.uint8(2)
+    yield np.uint16(2)
+    yield np.uint32(2)
+    yield np.uint64(2)
```
is there a reason np.bool8 not needed here or is it not necessary ?

seberg replied:
Thanks, seems I also forgot strings while refactoring...

seberg replied:
Argg, I commented it out for now, I had not thought of gh-9875 and my array-coercion PR of course homogenizes behaviour in that regard.  Although mainly for the numpy types, not for general strings.
We need to figure it out, either rip of the band-aid and fix it, or add on some type of band-aid before 1.20, for now I will open an issue for this and the bad float cases...

seberg commented:
Thought I remembered that there was something: https://docs.pytest.org/en/latest/example/parametrize.html#paramexamples  you can yield a `pytest.param` object which can have an `id`.

charris commented:
> Thought I remembered that there was something:

Looks like that should work.

seberg committed:
MAINT: Fix strings and use pytest.param to clean things up

seberg committed:
Merge branch 'new-coercion-tests' of github.com:seberg/numpy into new-coercion-tests

charris commented:
Looks much better. Strings need ids and looks like the nested parametrized cases in `TestTimeScalars` could also use some.

seberg committed:
TST: Add tests for 0-D array-like input to np.array() corner-cases

seberg committed:
TST: Improve parameterization IDs to get better printing at -vv

seberg commented:
@charris I think I improved the cases you were looking at, with the last commit?

charris commented:
Thanks Sebastian.
